model/stacktrace_frame.go

```diff
@@ -59,49 +59,43 @@ func (s *StacktraceFrame) Transform(service Service, smapAccessor utility.SmapAc
 	return m
 }
 
-func (s *StacktraceFrame) applySourcemap(service Service, smapAccessor utility.SmapAccessor) {
-	if service.Name == "" ||
-		service.Version == nil ||
-		s.AbsPath == nil ||
-		s.Colno == nil {
-		s.updateError("AbsPath, Colno, Service Name and Version mandatory for sourcemapping.")
+func (s *StacktraceFrame) applySourcemap(service Service, mapper sourcemap.Mapper) {
+	if s.Colno == nil {
+		s.updateError("Colno mandatory for sourcemapping.")
 		return
 	}
-
-	cleanedPath := utility.CleanUrlPath(*s.AbsPath)
-	smapConsumer, err := smapAccessor.Fetch(utility.SmapID{
-		ServiceName:    service.Name,
-		ServiceVersion: *service.Version,
-		Path:           cleanedPath,
-	})
+	sourcemapId := s.buildSourcemapId(service)
+	mapping, err := mapper.Apply(sourcemapId, s.Lineno, *s.Colno)
 	if err != nil {
 		logp.Err(fmt.Sprintf("Sourcemap fetching Error %s", err.Error()))
-		e, isSmapError := err.(utility.SmapError)
-		if !isSmapError || e.Kind == utility.MapError {
+		e, issourcemapError := err.(sourcemap.Error)
+		if !issourcemapError || e.Kind == sourcemap.MapError || e.Kind == sourcemap.KeyError {
```

Comment:
Is there a test to cover this?  if `!issourcemapError` `e` will be `nil` right?

Reply:
[This test](https://github.com/simitt/apm-server/blob/d5e687adf0935ddca571de56336a5de27174111f/model/stacktrace_frame_test.go#L100) covers it, see the [mocked return values](https://github.com/simitt/apm-server/blob/d5e687adf0935ddca571de56336a5de27174111f/model/stacktrace_frame_test.go#L218)

Reply:
As I am not using `e`, but `err` to retrieve the message I cannot see a problem here.

Reply:
Ugh, sorry, I was thinking the conditional wouldn't short-circuit, leaving `e` as `nil`, disregard.